store/src/commonMain/kotlin/com/dropbox/external/store5/impl/ShareableMarket.kt

```
@@ -0,0 +1,431 @@
+            }
+
+            if (response is MarketResponse.Success) {
+                stores.forEach { store -> (store.write as Write<Key, Output>).invoke(key, response.value) }
+
+                readCompletions.access {
+                    it[key]!!.forEach { anyOnCompletion ->
+                        val onCompletion = anyOnCompletion as OnMarketCompletion<Output>
+                        onCompletion.onSuccess(response)
+                    }
+                }
+            }
+
+            broadcasts.access {
+                scope.launch {
+                    it[key]!!.emit(response)
+                }
+            }
+        } catch (throwable: Throwable) {
+            val response = MarketResponse.Failure(
+                error = throwable,
+                origin = MarketResponse.Companion.Origin.Remote
+            )
+
+            broadcasts.access {
+                scope.launch {
+                    it[key]!!.emit(response)
+                }
+            }
+
+            readCompletions.access {
+                it[key]!!.forEach { anyOnCompletion ->
+                    val onCompletion = anyOnCompletion as OnMarketCompletion<Output>
+                    onCompletion.onFailure(response)
+                }
+            }
+        }
+    }
+
+    private suspend fun <Input : Any, Output : Any> tryPost(key: Key): RemoteResult<Input> {
+        return try {
+            val request = getLatestWriteRequest<Input, Output>(key)
+            val last = lastOrNull<Input>(key) ?: throw Exception()
+            return when (request.post(key, last)) {
+                else -> RemoteResult.Success(last)
+            }
+        } catch (throwable: Throwable) {
+            RemoteResult.Failure(throwable)
+        }
+    }
+
+    override suspend fun delete(key: Key): Boolean {
+        for (store in stores) {
+            try {
+                when (store.delete(key)) {
```

NIT: This can be rewritten as

```
if (!store.delete(key)) {
   return false
}
```

Couple questions:
1) Do we want to avoid any broadcast at all even if we were able to successfully delete a store?
2) Do we want to fail fast?

---

1. The broadcast must be active if this code is executed. My thought was we would only want `last()` to return `MarketResponse.Empty` if the delete was successful for each store and none have a value for that `Key`
2. Alternatively we would continue on failure to delete what we can? I think we should only emit `MarketResponse.Empty` if each store has no value for that `Key`. And so I think we can bail as soon as there is one failure